src/internal_modules/roc_sndio/target_pulseaudio/roc_sndio/pulseaudio_device.cpp

```diff
@@ -177,10 +177,19 @@ PulseaudioDevice::PulseaudioDevice(audio::FrameFactory& frame_factory,
 }
 
 PulseaudioDevice::~PulseaudioDevice() {
+    if (mainloop_) {
+        roc_panic("pulseaudio %s: device not closed", device_type_to_str(device_type_));
+    }
+}
+
+status::StatusCode PulseaudioDevice::close() {
     roc_log(LogDebug, "pulseaudio %s: closing device", device_type_to_str(device_type_));
 
-    close_();
-    stop_mainloop_();
+    if (close_() != status::StatusOK || stop_mainloop_() != status::StatusOK) {
+        return status::StatusErrDevice;
+    }
```

Comment:
Since PulseaudioDevice is designed so that close never fails, there is no need to return status codes from `close_()` and `stop_mainloop_()`.

For audio devices, closing just disconnects from device. Errors during closing are more useful for file devices, when close may write unbuffered data to disk and fail.